src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfoCacheService.cs

```diff
@@ -59,138 +91,124 @@ public SymbolTreeInfoCacheService()
+            tasks.Add(CreateWorkAsync(() => this.UpdateSourceSymbolTreeInfoAsync(project, cancellationToken), cancellationToken));
+
+            // Add tasks to update the symboltree for all metadata references.  As these are all distinct, they can run
+            // in parallel as we won't be trying to update the associated data for the same reference at the same time.
+            foreach (var reference in project.MetadataReferences.OfType<PortableExecutableReference>().Distinct())
+                tasks.Add(CreateWorkAsync(() => UpdateReferenceAsync(project, reference, cancellationToken), cancellationToken));
 
             await Task.WhenAll(tasks).ConfigureAwait(false);
         }
 
         private async Task UpdateSourceSymbolTreeInfoAsync(Project project, CancellationToken cancellationToken)
         {
-            var checksum = await SymbolTreeInfo.GetSourceSymbolsChecksumAsync(project, cancellationToken).ConfigureAwait(false);
+            // Find the top-level-semantic-version of this project.  We only want to recompute if it has changed. This is
+            // because the symboltree contains the names of the types/namespaces in the project and would not change
+            // if the semantic-version of the project hasn't changed.  We also do not need to check the 'dependent
+            // version'.  As this is just tracking parent/child relationships of namespace/type names for the source
+            // types in this project, this cannot change based on what happens in other projects.
+            var semanticVersion = await project.GetSemanticVersionAsync(cancellationToken).ConfigureAwait(false);
+
             if (!_projectIdToInfo.TryGetValue(project.Id, out var projectInfo) ||
-                projectInfo.Checksum != checksum)
+                projectInfo.semanticVersion != semanticVersion)
             {
-                projectInfo = await SymbolTreeInfo.GetInfoForSourceAssemblyAsync(
-                    project, checksum, cancellationToken).ConfigureAwait(false);
+                // If the checksum is the same (which can happen if we loaded the previous index from disk), then no
+                // need to recompute.
+                var checksum = await SymbolTreeInfo.GetSourceSymbolsChecksumAsync(project, cancellationToken).ConfigureAwait(false);
+                if (projectInfo.info?.Checksum != checksum)
+                {
+                    // Otherwise, looks like things changed.  Compute and persist the latest index.
+                    var info = await SymbolTreeInfo.GetInfoForSourceAssemblyAsync(
+                        project, checksum, cancellationToken).ConfigureAwait(false);
 
-                Contract.ThrowIfNull(projectInfo);
-                Contract.ThrowIfTrue(projectInfo.Checksum != checksum, "If we computed a SymbolTreeInfo, then its checksum much match our checksum.");
+                    Contract.ThrowIfNull(info);
```

Comment:
```suggestion
```

`GetInfoForSourceAssemblyAsync` appears to return a non-nullable

Reply:
yeah, but i like contracts for that for catching things :D